### PR TITLE
Fix calls to base implementation decompilation

### PR DESCRIPTION
**Pull request by magicmoux**
@hazzik,

Here is an up-to-date PR that should solve the calls to base method/properties implementations issue I talked about in #121 

I annotated the relevant changes so you don't get confused by all the code cleaning changes (sorry about these but I use CodeMaid on save)
Max.

### CODE REVIEW

File: src/DelegateDecompiler/MethodBodyDecompiler.cs [hunk 1]
```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
...
         {
+            HashSet<Tuple<object, MethodInfo>> baseCalls = new HashSet<Tuple<object, MethodInfo>>();
+
             var body = method.GetMethodBody();
             var addresses = new VariableInfo[body.LocalVariables.Count];
             for (var i = 0; i < addresses.Length; i++)
                 addresses[i] = new VariableInfo(body.LocalVariables[i].LocalType);
             var locals = addresses.ToArray();
 
+            if (!method.IsStatic && DelegateDecompiler.DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.First().DiscardConversion(), method)))
```
Comment (magicmoux) [inline src/DelegateDecompiler/MethodBodyDecompiler.cs]: Once a concrete implementation is to be decompiled we register any possible call to overridden member to be decompiled concretely for this particular instance.

File: src/DelegateDecompiler/MethodBodyDecompiler.cs [hunk 2]
```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
...
-                .Select(p => (Address) Expression.Parameter(p.ParameterType, p.Name))
+                .Select(p => (Address)Expression.Parameter(p.ParameterType, p.Name))
                 .ToList();
 
             var methodType = declaringType ?? method.DeclaringType;
             if (!method.IsStatic)
                 args.Insert(0, Expression.Parameter(methodType, "this"));
 
-            var expression = method.IsVirtual
+            var expression = method.IsVirtual && !DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.FirstOrDefault(), method))
```
Comment (magicmoux) [inline src/DelegateDecompiler/MethodBodyDecompiler.cs]: Here we also check whether a concrete call has been performed for this particular instance

File: src/DelegateDecompiler/Processor.cs [hunk 3]
```diff
@@ -1095,8 +1098,13 @@ static Expression[] GetArguments(ProcessorState state, MethodBase m)
             return mArgs;
         }
 
-        static Expression BuildMethodCallExpression(MethodInfo m, Address instance, Expression[] arguments)
+        private static Expression BuildMethodCallExpression(MethodInfo m, Address instance, Expression[] arguments)
         {
+            var concreteInstance = instance.DiscardConversion();
+            if (DelegateDecompiler.DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(concreteInstance, m)))
```
Comment (magicmoux) [inline src/DelegateDecompiler/Processor.cs]: Here we also check if a concrete call on the virtual chain has been performed for this particular instance

File: src/DelegateDecompiler/MethodBodyDecompiler.cs [hunk 4]
```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
...
             var body = method.GetMethodBody();
             var addresses = new VariableInfo[body.LocalVariables.Count];
             for (var i = 0; i < addresses.Length; i++)
                 addresses[i] = new VariableInfo(body.LocalVariables[i].LocalType);
             var locals = addresses.ToArray();
 
+            if (!method.IsStatic && DelegateDecompiler.DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.First().DiscardConversion(), method)))
+            {
+                var declaringType = method.DeclaringType;
+                baseCalls.UnionWith(AppDomain.CurrentDomain.GetAssemblies()
```
Comment (magicmoux) [inline src/DelegateDecompiler/MethodBodyDecompiler.cs]: Here we get every possible calls to base methods for the instance so they get decompiled concretely

File: src/DelegateDecompiler/MethodBodyDecompiler.cs [hunk 5]
```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
...
+
             var instructions = method.GetInstructions();
-            return Processor.Process(locals, args, instructions.First(), method.ReturnType);
+            try
+            {
+                return Processor.Process(locals, args, instructions.First(), method.ReturnType);
+            }
+            finally
+            {
+                foreach (var call in baseCalls)
```
Comment (magicmoux) [inline src/DelegateDecompiler/MethodBodyDecompiler.cs]: Once the concrete decompilation is complete, we clean up the registry for the current instance/method

**Comment by hazzik** [comment]
@magicmoux could you please explain what is the desired behaviour? I'm confused.

**Comment by magicmoux** [comment]
@hazzik what is the confusion about?

**Comment by hazzik** [comment]
@magicmoux I’ve made some changes on `develop` branch. Could you check if these changes work in your scenarios?

**Comment by magicmoux** [comment]
Hi @hazzik,

I've checked your implementation and I found out that it works only in one direction, ie. when you invoke the virtual call on a super-class. 
When called directly on a subclass, concrete calls to base won't be decompile since there are not added to the baseCalls dictionary.

You should have added the following code in MethodBodyDecompiler.cs @ before line 67 :

            var baseCalls = AppDomain.CurrentDomain.GetAssemblies()

                .Where(a => !a.IsDynamic)

                .SelectMany(a => SafeGetTypes(a))

                .Where(t => t.IsAssignableFrom(declaringType) && t != declaringType)

                .Select(t => GetDeclaredMethod(t, method))

                .Where(m => m!=null && !m.IsAbstract)

                .ToDictionary(m => m, m => DecompileConcrete(m, args));

**Comment by magicmoux** [comment]
On afterthought, that should not cause much overhead in most scenarii, but i didn't really check how the main cache works, so many decompilations could have négative impact on real environments

**Comment by magicmoux** [comment]
Hi, @hazzik 
still no concolusion on this problem?
I think this is one of the major issues that need to be resolved to reach a fully functional version.

**Pull request closed without merging**